bundles/org.openhab.binding.kaleidescape/src/main/java/org/openhab/binding/kaleidescape/internal/handler/KaleidescapeHandler.java

```diff
@@ -0,0 +1,617 @@
+            configError = "undefined serialPort and host configuration settings; please set one of them";
+        } else if (host == null || host.isEmpty()) {
+            if (serialPort != null && serialPort.toLowerCase().startsWith("rfc2217")) {
+                configError = "use host and port configuration settings for a serial over IP connection";
+            }
+        } else {
+            if (port == null) {
+                configError = "undefined port configuration setting";
+            } else if (port <= 0) {
+                configError = "invalid port configuration setting";
+            }
+        }
+
+        if (configError != null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, configError);
+            return;
+        }
+
+        if (componentType != null) {
+            this.componentType = componentType;
+        }
+
+        if (updatePeriod != null) {
+            this.updatePeriod = updatePeriod;
+        }
+
+        boolean channelsRemoved = false;
+        List<Channel> channels = new ArrayList<>(this.getThing().getChannels());
+
+        // check if volume is enabled, if not remove the volume & mute channels
+        if (config.volumeEnabled) {
+            this.volumeEnabled = true;
+            this.volume = config.initialVolume;
+            this.updateState(VOLUME, new PercentType(this.volume));
+            this.updateState(MUTE, OnOffType.OFF);
+        } else {
+            channelsRemoved = true;
+            channels.removeIf(c -> (c.getUID().getId().equals(VOLUME)));
+            channels.removeIf(c -> (c.getUID().getId().equals(MUTE)));
+        }
+
+        // remove music channels if we are not a Premiere Player or Cinema One
+        if (!(PLAYER.equals(this.componentType) || CINEMA_ONE.equals(this.componentType))) {
+            channelsRemoved = true;
+            channels.removeIf(c -> (c.getUID().getId().contains(MUSIC)));
+            channels.removeIf(c -> (c.getUID().getId().equals(DETAIL + DETAIL_ALBUM_TITLE)));
+            channels.removeIf(c -> (c.getUID().getId().equals(DETAIL + DETAIL_ARTIST)));
+            channels.removeIf(c -> (c.getUID().getId().equals(DETAIL + DETAIL_REVIEW)));
+        }
+
+        // premiere players do not support SYSTEM_READINESS_STATE
+        if (PLAYER.equals(this.componentType)) {
+            channelsRemoved = true;
+            channels.removeIf(c -> (c.getUID().getId().equals(SYSTEM_READINESS_STATE)));
+        }
+
+        // remove VIDEO_COLOR and CONTENT_COLOR if not a Strato
+        if (!STRATO.equals(this.componentType)) {
+            channelsRemoved = true;
+            channels.removeIf(c -> (c.getUID().getId().equals(VIDEO_COLOR)));
+            channels.removeIf(c -> (c.getUID().getId().equals(VIDEO_COLOR_EOTF)));
+            channels.removeIf(c -> (c.getUID().getId().equals(CONTENT_COLOR)));
+            channels.removeIf(c -> (c.getUID().getId().equals(CONTENT_COLOR_EOTF)));
+        }
```

Comment:
It would be cleaner to use separate thing definitions for these different models. The UI initially just reads the xml and then would show all channels even if not available if one would create the thing manually in the ui. You would not need to specify the component type as parameter. The xml will become a bit larger. Although you can use a config reference to point to a single configuration.

Reply:
Agreed, 4 things are now implemented.